cranelift/codegen/src/ir/memflags.rs

```diff
@@ -3,38 +3,74 @@
 use core::fmt;
 
 enum FlagBit {
+    // Flags to introduce a limited form of undefined behavior.
     Notrap,
     Aligned,
     Readonly,
+    // Flags to specify endianness of the memory access.  If BigEndian is
+    // present, the access is big-endian; otherwise, it is little-endian.
+    // This flag should always come last in the enumeration.
+    BigEndian,
 }
 
+// Note: The NAMES array only hold the names of flags for forms of undefined behavior
+// Endianness is handled separately.
 const NAMES: [&str; 3] = ["notrap", "aligned", "readonly"];
+const LITTLEENDIAN: &str = "little";
+const BIGENDIAN: &str = "big";
+
+/// Endianness of a memory access.
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+pub enum Endianness {
+    /// Little-endian
+    Little,
+    /// Big-endian
+    Big,
+}
 
 /// Flags for memory operations like load/store.
 ///
 /// Each of these flags introduce a limited form of undefined behavior. The flags each enable
 /// certain optimizations that need to make additional assumptions. Generally, the semantics of a
 /// program does not change when a flag is removed, but adding a flag will.
+///
+/// In addition, the flags determine the endianness of the memory access.  As opposed to the
+/// optimization flags defined above, modifying the endianness would always change program
+/// semantics, therefore the endianness must always be explicitly specified when constructing
+/// a MemFlags value, and cannot be changed later.
 #[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
 pub struct MemFlags {
     bits: u8,
 }
 
 impl MemFlags {
-    /// Create a new empty set of flags.
-    pub fn new() -> Self {
-        Self { bits: 0 }
+    /// Create a new set of flags.  This is initially empty except for the endianness.
+    /// Each MemFlags value always determines the endianness of the memory access.
+    pub fn new(endianness: Endianness) -> Self {
```

Comment:
I think using the native endian by default would be nicer. Except for WASM you almost always want to use the native endian.

Reply:
But where would you get that?   It depends on the target.

Reply:
An empty `MemFlags` would mean native endian, but you could add a flag to override it to either little endian or big endian. A little endian target will then look if the big endian flag exists and use little endian otherwise. A big endian target will look if the little endian flag exists and use big endian otherwise. You could add a method to `MemFlags` that accepts the native endian and computes the real endian of the memory instruction to avoid duplication of this logic.

Reply:
Also, the point of requiring *every* MemFlags to carry explicit endianness was to make it impossible to accidentally "lose" endianness e.g. during IR transformations.    This is of course a decision that can be made either way: is it more important to try and prevent this type of accidental bug, at the cost of added verbosity, or the other way round?

Reply:
I think this is a case where we want to prefer explicitness: otherwise it's too easy to introduce assumptions that break only on big-endian systems. Hopefully we eventually have such a system on CI, but it's still better to lift the safety to the API/type level.

Reply:
Also, to add to above: it might be nice to add some rationale here; surely some readers will be confused by this, as it feels a bit odd coming from e.g. LLVM. We should state that this arises from the goal to have a completely machine-independent IR, whose meaning doesn't change across targets; in contrast, other IRs often are generated with target information already in mind. The client that generates IR *may* wish to consult target info to optimally choose "native" endian for cases where it doesn't matter, but it's better to be explicit about this in all cases.